neqo-transport/src/send_stream.rs

```diff
@@ -1063,61 +1106,174 @@ impl ::std::fmt::Display for SendStream {
 }
 
 #[derive(Debug, Default)]
-pub(crate) struct SendStreams(IndexMap<StreamId, SendStream>);
+pub(crate) struct SendStreams {
+    map: IndexMap<StreamId, SendStream>,
+
+    // What we really want is a Priority Queue that we can do arbitrary
+    // removes from (so we can reprioritize). BinaryHeap doesn't work,
+    // because there's no remove().  BTreeMap doesn't work, since you can't
+    // duplicate keys.  PriorityQueue does have what we need, except for an
+    // ordered iterator that doesn't consume the queue.  So we roll our own.
+
+    // So we use a sorted Vec<> for the regular streams (that's usually all of
+    // them), and then a BTreeMap of an entry for each SendOrder value, and
+    // for each of those entries a HashSet of the stream_ids at that
+    // sendorder.  In most cases (such as stream-per-frame), there will be
+    // a single stream at a given sendorder.
+
+    // These both store stream_ids, which need to be looked up in 'map'.
+    // This avoids the complexity of trying to hold references to the
+    // Streams which are owned by the IndexMap.
+    sendordered: BTreeMap<SendOrder, HashSet<StreamId>>,
+    regular: Vec<StreamId>, // streams with no SendOrder set, sorted in stream_id order
```

Comment:
Possibly, but:  In stream-per-frame (and perhaps most instances where sendorder is used), all or almost all the streams will have sendorders, and so we'll iterate over map but ignore almost or all the iterations.   That seems a higher overhead (every time we're sending packets) than the maintenance overhead.

Reply:
Hmm. Having a count of the number of unordered streams would allow us to add `.take(count)` when iterating, which would shortcut very nicely. WDYT?

Reply:
I'm not sure I understand your suggestion here

Reply:
If, rather than maintaining a collection of streams that are unordered, you merely kept a count, then in the two cases we think most likely:

1. All streams are unordered.  You iterate over the main `IndexMap` directly, but `.take(x)` where `x = self.map.len()`, so you just iterate the map.
2. All streams are ordered.  You iterate over the main `IndexMap`, but `.take(0)`, which turns into a no-op.

The advantage is that you don't have an indirection for the first iteration; you iterate the collection directly.  In the second, you short-circuit that iteration (because iterators are lazy and `.take(0)` doesn't hit the underlying iterator at all.  It's only when you have a mix that you end up paying extra, but that's not a situation we think likely.  If there are a few streams that are unordered, then you'll have a few to track, but maybe you can count the number of unordered streams with unsent data so that you have to hit the collection less often.